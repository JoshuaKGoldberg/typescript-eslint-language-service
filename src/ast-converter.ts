// Note:
//
// Almost all line in this source file are copied from https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/parser.ts ,
// This code is workaround for https://github.com/typescript-eslint/typescript-eslint/issues/774 .

import ts from "typescript";
import { SourceCode, AST, Scope } from "eslint";
import { ParserOptions } from "@typescript-eslint/types";
import { analyze, AnalyzeOptions } from "@typescript-eslint/scope-manager";
import { visitorKeys } from "@typescript-eslint/visitor-keys";
import { ParseAndGenerateServicesResult, TSESTreeOptions } from "@typescript-eslint/typescript-estree";

import { sync as globSync } from "globby";
import isGlob from "is-glob";

/* The following types or functions are typescript-estree internals */
import { Extra } from "@typescript-eslint/typescript-estree/dist/parser-options";
import * as TsEstree from "@typescript-eslint/typescript-estree/dist/ast-converter";
import {
  CanonicalPath,
  getCanonicalFileName,
  ensureAbsolutePath,
} from "@typescript-eslint/typescript-estree/dist/create-program/shared";

function validateBoolean(value: boolean | undefined, fallback = false): boolean {
  if (typeof value !== "boolean") {
    return fallback;
  }
  return value;
}

function createExtra(code: string) {
  const base: Extra = {
    debugLevel: new Set(),
    tokens: null,
    range: true,
    loc: true,
    comment: false,
    comments: [],
    strict: false,
    jsx: false,
    useJSXTextNode: false,
    log: () => {},
    programs: null,
    projects: [],
    errorOnUnknownASTType: false,
    errorOnTypeScriptSyntacticAndSemanticIssues: false,
    code: "",
    tsconfigRootDir: process.cwd(),
    extraFileExtensions: [],
    preserveNodeMaps: undefined,
    createDefaultProgram: false,
    filePath: "",
    EXPERIMENTAL_useSourceOfProjectReferenceRedirect: false,
    singleRun: false,
    moduleResolver: "",
  };
  return {
    ...base,
    code,
    tokens: [],
    loc: true,
    comment: true,
    comments: [],
  };
}

function getTsconfigPath(tsconfigPath: string, extra: Extra): CanonicalPath {
  return getCanonicalFileName(ensureAbsolutePath(tsconfigPath, extra));
}

/**
 * Normalizes, sanitizes, resolves and filters the provided project paths
 */
function prepareAndTransformProjects(
  extra: Extra,
  projectsInput: string | string[] | undefined,
  ignoreListInput: string[],
): CanonicalPath[] {
  const sanitizedProjects: string[] = [];

  // Normalize and sanitize the project paths
  if (typeof projectsInput === "string") {
    sanitizedProjects.push(projectsInput);
  } else if (Array.isArray(projectsInput)) {
    for (const project of projectsInput) {
      if (typeof project === "string") {
        sanitizedProjects.push(project);
      }
    }
  }

  if (sanitizedProjects.length === 0) {
    return [];
  }

  // Transform glob patterns into paths
  const nonGlobProjects = sanitizedProjects.filter(project => !isGlob(project));
  const globProjects = sanitizedProjects.filter(project => isGlob(project));
  const uniqueCanonicalProjectPaths = new Set(
    nonGlobProjects
      .concat(
        globSync([...globProjects, ...ignoreListInput], {
          cwd: extra.tsconfigRootDir,
        }),
      )
      .map(project => getTsconfigPath(project, extra)),
  );

  // extra.log(
  //   'parserOptions.project (excluding ignored) matched projects: %s',
  //   uniqueCanonicalProjectPaths,
  // );

  return Array.from(uniqueCanonicalProjectPaths);
}

function applyParserOptionsToExtra(extra: Extra, options: TSESTreeOptions) {
  extra.range = typeof options.range === "boolean" ? options.range : extra.range;
  extra.loc = typeof options.loc === "boolean" ? options.loc : extra.range;
  /**
   * Track tokens in the AST
   */
  if (typeof options.tokens === "boolean" && options.tokens) {
    extra.tokens = [];
  }
  /**
   * Track comments in the AST
   */
  if (typeof options.comment === "boolean" && options.comment) {
    extra.comment = true;
    extra.comments = [];
  }
  /**
   * Enable JSX - note the applicable file extension is still required
   */
  if (typeof options.jsx === "boolean" && options.jsx) {
    extra.jsx = true;
  }
  /**
   * The JSX AST changed the node type for string literals
   * inside a JSX Element from `Literal` to `JSXText`.
   *
   * When value is `true`, these nodes will be parsed as type `JSXText`.
   * When value is `false`, these nodes will be parsed as type `Literal`.
   */
  if (typeof options.useJSXTextNode === "boolean" && options.useJSXTextNode) {
    extra.useJSXTextNode = true;
  }
  /**
   * Allow the user to cause the parser to error if it encounters an unknown AST Node Type
   * (used in testing)
   */
  if (typeof options.errorOnUnknownASTType === "boolean" && options.errorOnUnknownASTType) {
    extra.errorOnUnknownASTType = true;
  }
  /**
   * Allow the user to override the function used for logging
   */
  if (typeof options.loggerFn === "function") {
    extra.log = options.loggerFn;
  } else if (options.loggerFn === false) {
    extra.log = Function.prototype as any;
  }

  if (typeof options.tsconfigRootDir === "string") {
    extra.tsconfigRootDir = options.tsconfigRootDir;
  }

  // NOTE - ensureAbsolutePath relies upon having the correct tsconfigRootDir in extra
  extra.filePath = ensureAbsolutePath(extra.filePath, extra);

  if (Array.isArray(options.programs)) {
    if (options.programs.length) {
      extra.programs = options.programs;
    }
  }

  if (!extra.programs) {
    const projectFolderIgnoreList = (options.projectFolderIgnoreList ?? ["**/node_modules/**"])
      .reduce<string[]>((acc, folder) => {
        if (typeof folder === "string") {
          acc.push(folder);
        }
        return acc;
      }, [])
      // prefix with a ! for not match glob
      .map(folder => (folder.startsWith("!") ? folder : `!${folder}`));
    // NOTE - prepareAndTransformProjects relies upon having the correct tsconfigRootDir in extra
    extra.projects = prepareAndTransformProjects(extra, options.project, projectFolderIgnoreList);
  }

  if (Array.isArray(options.extraFileExtensions) && options.extraFileExtensions.every(ext => typeof ext === "string")) {
    extra.extraFileExtensions = options.extraFileExtensions;
  }

  /**
   * Allow the user to enable or disable the preservation of the AST node maps
   * during the conversion process.
   *
   * NOTE: For backwards compatibility we also preserve node maps in the case where `project` is set,
   * and `preserveNodeMaps` is not explicitly set to anything.
   */
  extra.preserveNodeMaps = typeof options.preserveNodeMaps === "boolean" && options.preserveNodeMaps;
  if (options.preserveNodeMaps === undefined && extra.projects.length > 0) {
    extra.preserveNodeMaps = true;
  }

  // extra.createDefaultProgram = typeof options.createDefaultProgram === "boolean" && options.createDefaultProgram;

  extra.EXPERIMENTAL_useSourceOfProjectReferenceRedirect =
    typeof options.EXPERIMENTAL_useSourceOfProjectReferenceRedirect === "boolean" &&
    options.EXPERIMENTAL_useSourceOfProjectReferenceRedirect;

  if (typeof options.moduleResolver === "string") {
    extra.moduleResolver = options.moduleResolver;
  }

  return extra;
}

export type AstConverterCreateOptions = {
  getProgram: () => ts.Program;
};

export class AstConverter {
  private readonly getProgram: () => ts.Program;

  public constructor({ getProgram }: AstConverterCreateOptions) {
    this.getProgram = getProgram;
  }

  /**
   *
   * see also https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/parser.ts#L346
   *
   */
  public parseAndGenerateServices(src: ts.SourceFile, options: TSESTreeOptions) {
    const code = src.getFullText();
    let extra: Extra = createExtra(code);

    extra.code = code;

    /**
     * Apply the given parser options
     */
    if (typeof options !== "undefined") {
      extra = applyParserOptionsToExtra(extra, options);
      if (
        typeof options.errorOnTypeScriptSyntacticAndSemanticIssues === "boolean" &&
        options.errorOnTypeScriptSyntacticAndSemanticIssues
      ) {
        extra.errorOnTypeScriptSyntacticAndSemanticIssues = true;
      }
    }

    const shouldProvideParserServices = extra.projects && extra.projects.length > 0;

    // Note: astConverter is an internal API
    const convert = TsEstree.astConverter;
    const { estree, astMaps } = convert(src, extra, true);

    /**
     * Return the converted AST and additional parser services
     */
    const ret: ParseAndGenerateServicesResult<any> = {
      ast: estree,
      services: {
        program: this.getProgram(),
        hasFullTypeInformation: shouldProvideParserServices,
        esTreeNodeToTSNodeMap: astMaps.esTreeNodeToTSNodeMap,
        tsNodeToESTreeNodeMap: astMaps.tsNodeToESTreeNodeMap,
      },
    };
    return ret;
  }

  /**
   * See also https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/parser/src/parser.ts
   */
  public parseForESLint(src: ts.SourceFile, options?: ParserOptions | null) {
    try {
      if (!options || typeof options !== "object") {
        options = {};
      }

      if (options.sourceType !== "module" && options.sourceType !== "script") {
        options.sourceType = "script";
      }
      if (typeof options.ecmaFeatures !== "object") {
        options.ecmaFeatures = {};
      }

      const parserOptions: TSESTreeOptions = {};
      Object.assign(parserOptions, options, {
        useJSXTextNode: validateBoolean(options.useJSXTextNode, true),
        jsx: validateBoolean(options.ecmaFeatures.jsx),
      });

      const analyzeOptions: AnalyzeOptions = {
        ecmaVersion: options.ecmaVersion,
        globalReturn: options.ecmaFeatures.globalReturn,
        jsxPragma: options.jsxPragma,
        jsxFragmentName: options.jsxFragmentName,
        lib: options.lib,
        sourceType: options.sourceType,
      };

      if (typeof options.filePath === "string") {
        const tsx = options.filePath.endsWith(".tsx");
        if (tsx || options.filePath.endsWith(".ts")) {
          parserOptions.jsx = tsx;
        }
      }

      /**
       * Allow the user to suppress the warning from typescript-estree if they are using an unsupported
       * version of TypeScript
       */
      const warnOnUnsupportedTypeScriptVersion = validateBoolean(options.warnOnUnsupportedTypeScriptVersion, true);
      if (!warnOnUnsupportedTypeScriptVersion) {
        parserOptions.loggerFn = false;
      }

      const { ast, services } = this.parseAndGenerateServices(src, parserOptions);
      ast.sourceType = options.sourceType;

      const scopeManager = analyze(ast, analyzeOptions);

      return {
        ast: ast as AST.Program,
        scopeManager: scopeManager as Scope.ScopeManager,
        services,
        visitorKeys: visitorKeys as SourceCode.VisitorKeys,
      };
    } catch (error) {
      throw error;
    }
  }

  public convertToESLintSourceCode(src: ts.SourceFile, options?: ParserOptions | null) {
    const code = src.getFullText();
    const { ast, scopeManager, services, visitorKeys } = this.parseForESLint(src, options);

    return new SourceCode({
      text: code,
      ast,
      scopeManager,
      parserServices: services,
      visitorKeys,
    });
  }
}
