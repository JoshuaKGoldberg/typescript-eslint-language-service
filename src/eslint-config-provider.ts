import path from "path";
import ts from "typescript";
import { Legacy } from "@eslint/eslintrc";
import type { ConfigArray, CascadingConfigArrayFactory } from "@eslint/eslintrc";

function getFactroyClass() {
  return Legacy.CascadingConfigArrayFactory;
}

export type ConfigProviderHost = {
  readonly readFile: (fileName: string, encoding: string) => string | undefined;
  readonly watchFile: (path: string, callback: ts.FileWatcherCallback, pollingInterval?: number) => ts.FileWatcher;
};

export type ESLintConfigProviderOptions = {
  host: ConfigProviderHost;
  log?: (msg: string) => void;
  directoriesToWatch: string[];
};

const ESLINTRC_SUFFIX_LIST = [
  ".eslintrc.js",
  ".eslintrc.yaml",
  ".eslintrc.yml",
  ".eslintrc.json",
  ".eslintrc",
  "package.json",
];

export interface ConfigProvider {
  getConfigArrayForFile(fileName: string): ConfigArray;
}

export class ESLintConfigProvider implements ConfigProvider {
  private readonly host: ConfigProviderHost;
  private readonly factory: CascadingConfigArrayFactory;
  private readonly log: (msg: string) => void;

  public constructor({ host, log = () => {}, directoriesToWatch }: ESLintConfigProviderOptions) {
    this.host = host;
    this.log = log;
    const eslintRecommendedPath = this.resolveESLintIntrinsicConfigPath("eslint-recommended");
    const eslintAllPath = this.resolveESLintIntrinsicConfigPath("eslint-all");

    //
    // It requires the ESLint intrinsic configuration js file path when user config has "eslint:recommended" or "eslint:all" such as:
    //
    // ```yaml
    // extends:
    //   - "eslint:recommended"
    // ```
    //
    this.factory = new (getFactroyClass())({
      eslintAllPath,
      eslintRecommendedPath,
    });

    directoriesToWatch.forEach(directory => {
      ESLINTRC_SUFFIX_LIST.map(suffix => path.resolve(directory, suffix)).forEach(eslintrcFilepath => {
        this.host.watchFile(eslintrcFilepath, () => this.factory.clearCache(), 50);
      });
    });
  }

  public getConfigArrayForFile(fileName: string) {
    return this.factory.getConfigArrayForFile(fileName);
  }

  public getConfigForFile(fileName: string) {
    return this.factory.getConfigArrayForFile(fileName).extractConfig(fileName);
  }

  private resolveESLintIntrinsicConfigPath(name: "eslint-all" | "eslint-recommended") {
    let ret: string | undefined = undefined;
    try {
      const fragments = require.resolve("eslint").split("node_modules/eslint");
      ret = [...fragments.slice(0, fragments.length - 1), `/conf/${name}.js`].join("node_modules/eslint");
    } catch (e: any) {
      this.log(e);
    }
    return ret;
  }
}
