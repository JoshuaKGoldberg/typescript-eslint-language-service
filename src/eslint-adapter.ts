import readPkgUp = require("read-pkg-up");
import ts from "typescript";
import { Linter, CLIEngine } from "eslint";
import { AstConverter } from "./ast-converter";
import { InvalidParserError } from "./errors";
import { ConfigProvider } from "./eslint-config-provider";
import { TS_LANGSERVICE_ESLINT_DIAGNOSTIC_ERROR_CODE } from "./consts";
import { ParserOptions } from "@typescript-eslint/parser";

export function translateToDiagnosticsFromESLintResult(
  result: Linter.LintMessage[],
  sourceFile: ts.SourceFile,
): ts.Diagnostic[] {
  return result.map(({ message, severity, ruleId, line, column, endLine, endColumn }) => {
    const messageText = `[${ruleId}] ${message}`;

    const category: ts.DiagnosticCategory =
      severity === 2
        ? ts.DiagnosticCategory.Error
        : severity === 1
        ? ts.DiagnosticCategory.Warning
        : ts.DiagnosticCategory.Suggestion;

    /**
     * ESLint uses 1-started index. On the other hand, TypeScript 0-started index.
     * So we should minus from ESLint result's row/column to create TypeScript position.
     */
    const start = ts.getPositionOfLineAndCharacter(sourceFile, line - 1, column - 1);
    const end = endLine && endColumn ? ts.getPositionOfLineAndCharacter(sourceFile, endLine - 1, endColumn - 1) : start;
    const length = Math.max(0, end - start);

    const diagnostic: ts.Diagnostic = {
      file: sourceFile,
      category,
      code: TS_LANGSERVICE_ESLINT_DIAGNOSTIC_ERROR_CODE,
      messageText,
      start,
      length,
    };

    return diagnostic;
  });
}

export function isIntersect(
  message: Linter.LintMessage,
  range: { start: number; end: number },
  sourceFile: ts.SourceFile,
) {
  const { line, column, endLine, endColumn } = message;
  const mStart = ts.getPositionOfLineAndCharacter(sourceFile, line - 1, column - 1);
  const mEnd = endLine && endColumn ? ts.getPositionOfLineAndCharacter(sourceFile, endLine - 1, endColumn - 1) : mStart;
  return !(mEnd < range.start || mStart > range.end);
}

export function translateToCodeFixesFromESLintResult(
  result: Linter.LintMessage[],
  fileName: string,
): ts.CodeFixAction[] {
  return result.reduce((acc, { message, ruleId, fix }) => {
    if (!fix) {
      return acc;
    }
    const rid = ruleId || "eslint";
    const rangeStart = fix.range[0];
    const rangeLength = fix.range[1] ? fix.range[1] - fix.range[0] : 0;
    const codeFixAction: ts.CodeFixAction = {
      description: `Fix: ${message}`,
      fixId: rid,
      fixName: rid,
      changes: [
        {
          fileName,
          isNewFile: false,
          textChanges: [
            {
              span: {
                start: rangeStart,
                length: rangeLength,
              },
              newText: fix.text,
            },
          ],
        },
      ],
    };
    return [...acc, codeFixAction];
  }, [] as ts.CodeFixAction[]);
}

export type ESLintAdapterOptions = {
  logger: (msg: string) => void;
  getSourceFile: (fileName: string) => ts.SourceFile | undefined;
  converter: AstConverter;
  configProvider: ConfigProvider;
};

export class ESLintAdapter {
  private readonly linter: Linter;
  private readonly logger: (msg: string) => void;
  private readonly converter: AstConverter;
  private readonly configProvider: ConfigProvider;
  private readonly getSourceFile: (fileName: string) => ts.SourceFile | undefined;

  public constructor({ logger, converter, configProvider, getSourceFile }: ESLintAdapterOptions) {
    this.linter = new Linter();
    this.logger = logger;
    this.converter = converter;
    this.configProvider = configProvider;
    this.getSourceFile = getSourceFile;
  }

  private getESLintResult(fileName: string, sourceFile: ts.SourceFile) {
    if (new CLIEngine({}).isPathIgnored(fileName)) {
      return [];
    }

    const configArray = this.configProvider.getConfigArrayForFile(fileName);
    const configFileContent = configArray.extractConfig(fileName).toCompatibleObjectAsConfigFileContent();
    if (
      !configFileContent.parser ||
      (readPkgUp.sync({ cwd: configFileContent.parser })?.packageJson.name ?? "") !== "@typescript-eslint/parser"
    ) {
      throw new InvalidParserError();
    }
    const parserOptions = (configFileContent.parserOptions ? configFileContent.parserOptions : {}) as ParserOptions;
    const sourceCode = this.converter.convertToESLintSourceCode(sourceFile, parserOptions);

    // See https://github.com/eslint/eslint/blob/v6.1.0/lib/linter/linter.js#L1130
    return this.linter.verify(sourceCode, configArray as any, { filename: fileName });
  }

  public getSemanticDiagnostics(
    delegate: ts.LanguageService["getSemanticDiagnostics"],
    fileName: string,
  ): ReturnType<ts.LanguageService["getSemanticDiagnostics"]> {
    const original = delegate(fileName);
    try {
      const sourceFile = this.getSourceFile(fileName);
      if (!sourceFile) {
        return original;
      }
      const eslintResult = this.getESLintResult(fileName, sourceFile);

      return [...original, ...translateToDiagnosticsFromESLintResult(eslintResult, sourceFile)];
    } catch (error) {
      if (error instanceof Error) {
        this.logger(error.message);
        if (error.stack) {
          this.logger(error.stack);
        }
      } else {
        this.logger(error);
      }
      return original;
    }
  }

  public getCodeFixesAtPosition(
    delegate: ts.LanguageService["getCodeFixesAtPosition"],
    fileName: string,
    start: number,
    end: number,
    errorCodes: number[],
    formatOptions: ts.FormatCodeSettings,
    preferences: ts.UserPreferences,
  ): ReturnType<ts.LanguageService["getCodeFixesAtPosition"]> {
    const original = delegate(fileName, start, end, errorCodes, formatOptions, preferences);
    try {
      if (!errorCodes.includes(TS_LANGSERVICE_ESLINT_DIAGNOSTIC_ERROR_CODE)) {
        return original;
      }

      const sourceFile = this.getSourceFile(fileName);
      if (!sourceFile) {
        return original;
      }

      const eslintResult = this.getESLintResult(fileName, sourceFile);

      return [
        ...original,
        ...translateToCodeFixesFromESLintResult(
          eslintResult.filter(r => isIntersect(r, { start, end }, sourceFile)),
          fileName,
        ),
      ];
    } catch (error) {
      this.logger(error.message ? error.message : "unknow error");
      return original;
    }
    return original;
  }
}
