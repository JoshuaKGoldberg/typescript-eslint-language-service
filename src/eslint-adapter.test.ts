import path from "path";
import ts from "typescript";
import { mark, Frets } from "fretted-strings";
import { AstConverter } from "./ast-converter";
import { ESLintAdapter } from "./eslint-adapter";
import { ConfigProvider } from "./eslint-config-provider";
import { ConfigArray } from "eslint/lib/cli-engine/config-array/config-array";
import { TS_LANGSERVICE_ESLINT_DIAGNOSTIC_ERROR_CODE } from "./consts";

class TestingConfigProvider implements ConfigProvider {
  public conf: any = {
    parser: {
      error: null,
      id: "@typescript-eslint/parser",
      filePath: path.resolve(__dirname, "../node_modules/@typescript-eslint/parser/dist/parser.js"),
    },
  };
  public getConfigArrayForFile(): ConfigArray {
    return new ConfigArray(this.conf);
  }
}

// `diagnostic.file` is so noisy for snapshot test
function filterSourceFileFromDiagnosticList(diagnostics: ts.Diagnostic[]) {
  return diagnostics.map(d => {
    delete d.file;
    return d;
  });
}

const getProgram = () => ({ } as any as ts.Program);

describe("ESLintAdapter", () => {
  describe("#getSemanticDiagnostics", () => {
    it("shuld return ESLint verification result as TypeScript diagnostic format", () => {
      const configProvider = new TestingConfigProvider();
      configProvider.conf.rules = {
        semi: 2,
      };
      const adapter = new ESLintAdapter({
        converter: new AstConverter({
          getProgram,
        }),
        getSourceFile: () => ts.createSourceFile("main.ts", "const x = 1", ts.ScriptTarget.ESNext, true),
        configProvider,
        logger: () => { },
      });
      const diagnostics = filterSourceFileFromDiagnosticList(adapter.getSemanticDiagnostics(() => [], "main.ts"));
      expect(diagnostics).toMatchSnapshot();
    });
  });

  describe("#getCodeFixesAtPosition", () => {

    it("shuld only delegate when input errorCodes does not include TS_LANGSERVICE_ESLINT_DIAGNOSTIC_ERROR_CODE", () => {
      const configProvider = new TestingConfigProvider();
      configProvider.conf.rules = {
        quotes: [2, "double"],
      };
      const frets: Frets = {};
      const content = mark(
        `
          'use strict';     
     %%%  ^           ^   %%%
     %%%  p1          p2  %%%
        `,
        frets,
      );
      const adapter = new ESLintAdapter({
        converter: new AstConverter({
          getProgram,
        }),
        getSourceFile: () => ts.createSourceFile("main.ts", content, ts.ScriptTarget.ESNext, true),
        configProvider,
        logger: () => { },
      });
      const codeFixes = adapter.getCodeFixesAtPosition(() => [], "main.ts", frets.p1.pos, frets.p2.pos, [], { }, { });
      expect(codeFixes).toMatchSnapshot();
    });

    it("shuld only delegate when input range isn't intersected ESLint reported results", () => {
      const configProvider = new TestingConfigProvider();
      configProvider.conf.rules = {
        quotes: [2, "double"],
      };
      const frets: Frets = {};
      const content = mark(
        `
          'use strict';     

          const hoge = 1;
     %%%  ^             ^   %%%
     %%%  p1            p2  %%%
        `,
        frets,
      );
      const adapter = new ESLintAdapter({
        converter: new AstConverter({
          getProgram,
        }),
        getSourceFile: () => ts.createSourceFile("main.ts", content, ts.ScriptTarget.ESNext, true),
        configProvider,
        logger: () => { },
      });
      const codeFixes = adapter.getCodeFixesAtPosition(() => [], "main.ts", frets.p1.pos, frets.p2.pos, [TS_LANGSERVICE_ESLINT_DIAGNOSTIC_ERROR_CODE], { }, { });
      expect(codeFixes).toMatchSnapshot();
    });

    it("shuld return ESLint sourceCodeFixer result as TypeScript codeFixAction format", () => {
      const configProvider = new TestingConfigProvider();
      configProvider.conf.rules = {
        quotes: [2, "double"],
      };
      const frets: Frets = {};
      const content = mark(
        `
          'use strict';     
     %%%  ^           ^   %%%
     %%%  p1          p2  %%%
        `,
        frets,
      );
      const adapter = new ESLintAdapter({
        converter: new AstConverter({
          getProgram,
        }),
        getSourceFile: () => ts.createSourceFile("main.ts", content, ts.ScriptTarget.ESNext, true),
        configProvider,
        logger: () => { },
      });
      const codeFixes = adapter.getCodeFixesAtPosition(() => [], "main.ts", frets.p1.pos, frets.p2.pos, [TS_LANGSERVICE_ESLINT_DIAGNOSTIC_ERROR_CODE], { }, { });
      expect(codeFixes).toMatchSnapshot();
    });
  });
});
