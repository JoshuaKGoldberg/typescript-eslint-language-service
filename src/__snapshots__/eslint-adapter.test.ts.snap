// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ESLintAdapter #getCodeFixesAtPosition shuld only delegate when input errorCodes does not include TS_LANGSERVICE_ESLINT_DIAGNOSTIC_ERROR_CODE 1`] = `[]`;

exports[`ESLintAdapter #getCodeFixesAtPosition shuld only delegate when input range isn't intersected ESLint reported results 1`] = `[]`;

exports[`ESLintAdapter #getCodeFixesAtPosition shuld return ESLint sourceCodeFixer result as TypeScript codeFixAction format 1`] = `
[
  {
    "changes": [
      {
        "fileName": "main.ts",
        "isNewFile": false,
        "textChanges": [
          {
            "newText": ""use strict"",
            "span": {
              "length": 12,
              "start": 11,
            },
          },
        ],
      },
    ],
    "description": "Fix: Strings must use doublequote.",
    "fixId": "quotes",
    "fixName": "quotes",
  },
]
`;

exports[`ESLintAdapter #getSemanticDiagnostics shuld return ESLint verification result as TypeScript diagnostic format 1`] = `
[
  {
    "category": 1,
    "code": 30010,
    "length": 0,
    "messageText": "[semi] Missing semicolon.",
    "start": 11,
  },
]
`;
